File: 0062-unique-paths/0062-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        grid = [[0] * n for _  in range(m)]
        
        grid[0][0] = 1
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if i == 0 or j == 0:
                    grid[i][j] = 1
                    continue
                grid[i][j] = grid[i-1][j] + grid[i][j-1]
                
                
        # print(grid)
                
        return grid[m-1][n-1]
                
        
#         if m == 1 and n == 1:
#             return 1
        
#         def move(row , col):
#             if row >= m or col >= n:
#                 return 0
#             if row == m-1 and col == n-1:
        
#                 return 1
#             if grid[row][col]:
#                 return grid[row][col]
#             grid[row][col] = move(row , col+1) + move(row+1 , col)
#             return grid[row][col]
            
            
            
#         move(0 , 0)
#         print(grid)
    
        # return grid[0][0]
                

    

File: 0062-unique-paths/0062-unique-paths.cpp
class Solution {
public:
    
    int count(int r , int c , vector<vector<int>>&dp){
        if(dp[r][c] != -1) return dp[r][c];
        
        if(r == 0 || c == 0) return 1;
        if(r < 0 || c < 0) return 0;
        int up = count(r-1 ,c , dp);
        int left = count(r , c-1 , dp);
        dp[r][c]= up + left;
        return dp[r][c] ;
    }
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m , vector<int>(n , -1));
        // dp[0][0] = 1;
        return count(m -1, n-1 , dp);
//         for(int i=1; i<m; i++){
//             for(int j = 1; j<n; j++){
//                 dp[i][j] = dp[i-1][j] + dp[i][j-1];
//             }
//         }
        
//         for(int i=0; i<m; i++){
//             for(int j=0; j<n; j++){
//                 cout<< "The values are: " << dp[i][j] << endl;
//             }
//         }
        // return dp[m-1][n-1];
        
    }
};